File: ProjectEuler/Problems/Problem026.cs
using Microsoft.Extensions.Logging;
using System.Collections.Generic;

namespace ProjectEuler.Problems
{
    public class Problem026 : Problem<int>
    {
        public Problem026(ILogger<Problem026> logger) : base(logger) { }

        protected override int CalculateAnswer()
        {
            var answer = 0;
            var n = 1000;
            var longestCycleLength = 0;

            for (var d = 2; d < n; ++d)
            {
                var cycleLength = CycleLength(d);
                if (cycleLength > longestCycleLength)
                {
                    answer = d;
                    longestCycleLength = cycleLength;
                    Logger.LogDebug("d = {d}; cycle length = {cycleLength}", d, cycleLength);
                }
            }

            return answer;
        }

        private static int MinPower10MultipleOfRGreaterThanD(int d, int r)
        {
            var a = r;
            while (d > a)
            {
                a *= 10;
            }
            return a;
        }

        private static int CycleLength(int d)
        {
            var result = 0;
            var r = 1;
            var seen = new Dictionary<int, int>();
            while (!seen.ContainsKey(r))
            {
                seen[r] = result;
                var dividend = MinPower10MultipleOfRGreaterThanD(d, r);
                r = dividend % d;
                if (r == 0) return 0;
                ++result;
            }
            return result - seen[r];
        }
    }
}

File: ProjectEulerPython/problems/problem_026.py
from problems.problem import Problem

def min_pow_10_mult_of_r_greater_than_d(d: int, r: int) -> int:
  a = r
  while d > a:
    a *= 10
  return a

def cycle_length(d: int) -> int:
  result = 0
  r = 1
  seen = {}
  while not r in seen:
    seen[r] = result
    dividend = min_pow_10_mult_of_r_greater_than_d(d, r)
    r = dividend % d
    if r == 0:
      return 0
    result += 1

  return result - seen[r]

class Problem026(Problem):

  def calculate_answer(self) -> int:
    answer = 0
    n = 1000
    longest_cycle_length = 0

    for d in range(2, n):
      length = cycle_length(d)
      if length > longest_cycle_length:
        answer = d
        longest_cycle_length = length
        self.print_detail("d = " + str(d) + "; cycle length = " + str(length))

    return answer